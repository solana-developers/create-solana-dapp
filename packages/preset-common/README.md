# @solana-developers/preset-common

This library is a preset for [create-solana-dapp](https://npm.im/create-solana-dapp) with shared code for the other
presets.

## Supported generators

### common-template

```bash
>  NX   generate @solana-developers/preset-common:common-template [name] [options,...]

From:  @solana-developers/preset-common (v0.0.1)
Name:  common-template

  Generate Common templates

Options:
    --directory                                                        [string]
    --name                                                             [string]
    --npmScope         The npm scope to use                            [string]
    --template         The template to use                   [string] [choices:
                                                           "license", "readme"]
    --webName          Name of the web application                     [string]
                       (overrides name)
    --anchor           Anchor template to use        [string] [choices: "none",
                                                      "counter", "hello-world"]
                                                              [default: "none"]
    --anchorName       Anchor project name         [string] [default: "anchor"]
    --licenseAuthor    The author to use in the                        [string]
                       license
    --preset           The preset to use             [string] [choices: "next",
                                                                       "react"]
```

## More information

- [GitHub repository](https://github.com/solana-developers/create-solana-dapp)
- [npm package](https://npm.im/create-solana-dapp)
- [StackOverflow](https://solana.stackexchange.com/questions/tagged/create-solana-dapp)
