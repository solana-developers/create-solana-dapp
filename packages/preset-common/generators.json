{
  "generators": {
    "common-template": {
      "factory": "./src/generators/common-template/common-template-generator",
      "schema": "./src/generators/common-template/common-template-schema.json",
      "description": "common-template generator"
    }
  }
}
