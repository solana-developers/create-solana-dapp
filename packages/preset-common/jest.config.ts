/* eslint-disable */
export default {
  displayName: 'preset-common',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\.[tj]s$': ['ts-jest', { tsconfig: '<rootDir>/tsconfig.spec.json' }],
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../coverage/packages/preset-common',
  testEnvironment: 'node',
}
