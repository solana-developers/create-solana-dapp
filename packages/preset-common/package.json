{
  "name": "@solana-developers/preset-common",
  "version": "0.0.1",
  "description": "This library is a preset for create-solana-dapp with shared code for the other presets.",
  "dependencies": {
    "@nx/devkit": "17.1.3",
    "tslib": "^2.3.0"
  },
  "type": "commonjs",
  "main": "./src/index.js",
  "typings": "./src/index.d.ts",
  "generators": "./generators.json",
  "keywords": [
    "solana",
    "next",
    "react",
    "vue",
    "web3",
    "yarn",
    "blockchain",
    "nft"
  ],
  "license": "MIT",
  "homepage": "https://github.com/solana-developers/create-solana-dapp#readme",
  "bugs": {
    "url": "https://github.com/solana-developers/create-solana-dapp/issues"
  },
  "repository": {
    "name": "solana-developers/create-solana-dapp",
    "type": "git",
    "url": "https://github.com/solana-developers/create-solana-dapp"
  },
  "contributors": [
    {
      "name": "Joe Caulfield",
      "url": "https://github.com/jpcaulfi"
    },
    {
      "name": "Jacob Creech",
      "url": "https://github.com/jacobcreech"
    },
    {
      "name": "Nick Frostbutter",
      "url": "https://github.com/nickfrosty"
    },
    {
      "name": "Bram Borggreve",
      "url": "https://github.com/beeman"
    }
  ]
}
