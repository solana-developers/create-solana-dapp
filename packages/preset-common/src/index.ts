export * from './generators/common-template/common-template-generator'
export * from './generators/common-template/common-template-schema'
export * from './utils/generic-substitutions'
export * from './utils/application-cleanup'
export * from './utils/get-recursive-file-contents'
export * from './utils/package-versions'
export * from './utils/run-command'
export * from './utils/update-ignore-entires'
