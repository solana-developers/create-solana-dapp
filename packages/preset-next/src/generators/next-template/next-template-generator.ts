import { generateFiles, Tree } from '@nx/devkit'
import * as path from 'path'
import { nextApplicationSubstitutions } from '../../utils'
import { NextTemplateSchema } from './next-template-schema'

export async function nextTemplateGenerator(tree: Tree, options: NextTemplateSchema) {
  const substitutions = nextApplicationSubstitutions({
    anchor: options.anchor,
    anchorName: options.anchorName,
    name: options.webName,
    npmScope: options.npmScope,
  })
  generateFiles(tree, path.join(__dirname, 'files', options.template), options.directory, {
    ...options,
    ...substitutions,
    fileNameUnderscore: substitutions.fileName.replace(/-/g, '_'),
  })
}

export default nextTemplateGenerator
