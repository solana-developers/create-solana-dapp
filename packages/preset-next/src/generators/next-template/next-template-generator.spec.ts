import { Tree } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import { getRecursiveFileContents } from '@solana-developers/preset-common'

import { nextTemplateGenerator } from './next-template-generator'
import { NextTemplateSchema } from './next-template-schema'

describe('next-template generator', () => {
  let tree: Tree
  const options: NextTemplateSchema = {
    name: 'test',
    template: 'base',
    directory: 'test',
    anchor: 'none',
    anchorName: 'anchor',
    npmScope: 'test',
    webName: 'test',
  }
  const templates: NextTemplateSchema['template'][] = ['base', 'none', 'tailwind']

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace()
  })

  it.each(templates)('should generate files for %s template', async (template) => {
    await nextTemplateGenerator(tree, { ...options, template })

    const contents = getRecursiveFileContents(tree, options.directory)
    expect(contents).toMatchSnapshot()
  })
})
