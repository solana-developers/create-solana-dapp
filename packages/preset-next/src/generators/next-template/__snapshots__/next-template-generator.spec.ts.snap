// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`next-template generator should generate files for base template 1`] = `
{
  "public": {
    "children": {
      "favicon.ico": {
        "content": null,
        "isBinary": true,
        "path": "test/public/favicon.ico",
      },
      "solana-logo.png": {
        "content": null,
        "isBinary": true,
        "path": "test/public/solana-logo.png",
      },
    },
    "path": "test/public",
  },
}
`;

exports[`next-template generator should generate files for none template 1`] = `
{
  "app": {
    "children": {
      "app-layout.tsx": {
        "content": [
          ""use client"",
          "import { ReactNode } from 'react'",
          "import Link from 'next/link'",
          "import dynamic from "next/dynamic";",
          "const WalletMultiButton = dynamic(",
          "async () =>",
          "(await import("@solana/wallet-adapter-react-ui")).WalletMultiButton,",
          "{ ssr: false }",
          ")",
          "export function AppLayout({ children }: { children: ReactNode }) {",
          "const { pathname } = { pathname: '' }",
          "const pages = [",
          "{ label: 'Airdrop', path: '/airdrop' },",
          "{ label: 'Page 1', path: '/page-1' },",
          "{ label: 'Page 2', path: '/page-2' },",
          "]",
          "return (",
          "<div>",
          "<div>",
          "<div>",
          "<Link href="/" >",
          "@solana-developers/preset-next",
          "</Link>",
          "<ul >",
          "{pages.map(({ label, path }) => (",
          "<li key={path}>",
          "<Link href={path}>",
          "{label}",
          "</Link>",
          "</li>",
          "))}",
          "</ul>",
          "</div>",
          "<div>",
          "<WalletMultiButton />",
          "</div>",
          "</div>",
          "<div>{children}</div>",
          "<footer>",
          "<aside>",
          "<p>Generated with @solana-developers/preset-next@0.0.1</p>",
          "</aside>",
          "</footer>",
          "</div>",
          ")",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/app-layout.tsx",
      },
      "global.css": {
        "content": [
          "html,",
          "body {",
          "height: 100%;",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/global.css",
      },
      "layout.tsx": {
        "content": [
          "import './global.css'",
          "import { AppLayout } from './app-layout'",
          "import { SolanaProvider } from './solana-provider'",
          "export const metadata = {",
          "title: 'Welcome to test',",
          "description: 'Generated by @solana-developers/preset-next@0.0.1',",
          "}",
          "export default function RootLayout({ children }: { children: React.ReactNode }) {",
          "return (",
          "<html lang="en">",
          "<body>",
          "<SolanaProvider>",
          "<AppLayout>",
          "{children}",
          "</AppLayout>",
          "</SolanaProvider>",
          "</body>",
          "</html>",
          ")",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/layout.tsx",
      },
      "page.tsx": {
        "content": [
          "export default async function Index() {",
          "return (",
          "<div>",
          "<div>Index</div>",
          "</div>",
          ")",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/page.tsx",
      },
      "solana-provider.tsx": {
        "content": [
          ""use client"",
          "import { WalletError } from '@solana/wallet-adapter-base';",
          "import {",
          "ConnectionProvider,",
          "WalletProvider,",
          "} from '@solana/wallet-adapter-react';",
          "import { WalletModalProvider } from '@solana/wallet-adapter-react-ui';",
          "import { SolflareWalletAdapter } from '@solana/wallet-adapter-solflare';",
          "import { UnsafeBurnerWalletAdapter } from '@solana/wallet-adapter-unsafe-burner';",
          "import { Cluster, clusterApiUrl } from '@solana/web3.js';",
          "import {",
          "createContext,",
          "ReactNode,",
          "useCallback,",
          "useContext,",
          "useMemo,",
          "useState,",
          "} from 'react';",
          "require('@solana/wallet-adapter-react-ui/styles.css');",
          "export interface SolanaProviderContext {",
          "cluster: Cluster;",
          "endpoint: string;",
          "setCluster?: (cluster: Cluster) => void;",
          "}",
          "const Context = createContext<SolanaProviderContext>(",
          "{} as SolanaProviderContext",
          ");",
          "export function SolanaProvider({ children }: { children: ReactNode }) {",
          "const [cluster, setCluster] = useState<Cluster>('devnet');",
          "const endpoint = useMemo(() => clusterApiUrl(cluster), [cluster]);",
          "const wallets = useMemo(",
          "() => [new UnsafeBurnerWalletAdapter(), new SolflareWalletAdapter()],",
          "[cluster]",
          ");",
          "const onError = useCallback((error: WalletError) => {",
          "console.error(error);",
          "}, []);",
          "const value: SolanaProviderContext = {",
          "cluster,",
          "endpoint,",
          "setCluster,",
          "};",
          "return (",
          "<ConnectionProvider endpoint={endpoint}>",
          "<WalletProvider wallets={wallets} onError={onError} autoConnect={true}>",
          "<WalletModalProvider>",
          "<Context.Provider value={value}>{children}</Context.Provider>",
          "</WalletModalProvider>",
          "</WalletProvider>",
          "</ConnectionProvider>",
          ");",
          "}",
          "export function useSolana() {",
          "return useContext(Context);",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/solana-provider.tsx",
      },
    },
    "path": "test/app",
  },
}
`;

exports[`next-template generator should generate files for tailwind template 1`] = `
{
  "app": {
    "children": {
      "app-layout.tsx": {
        "content": [
          ""use client"",
          "import { ReactNode } from 'react'",
          "import Link from 'next/link'",
          "import dynamic from "next/dynamic";",
          "const WalletMultiButton = dynamic(",
          "async () =>",
          "(await import("@solana/wallet-adapter-react-ui")).WalletMultiButton,",
          "{ ssr: false }",
          ")",
          "export function AppLayout({ children }: { children: ReactNode }) {",
          "const { pathname } = { pathname: '' }",
          "const pages = [",
          "{ label: 'Airdrop', path: '/airdrop' },",
          "{ label: 'Page 1', path: '/page-1' },",
          "{ label: 'Page 2', path: '/page-2' },",
          "]",
          "return (",
          "<div className="h-full flex flex-col">",
          "<div className="navbar bg-base-300 text-neutral-content">",
          "<div className="flex-1">",
          "<Link href="/" className="btn btn-ghost normal-case text-xl">",
          "@solana-developers/preset-next",
          "</Link>",
          "<ul className="menu menu-horizontal px-1">",
          "{pages.map(({ label, path }) => (",
          "<li key={path}>",
          "<Link className={pathname.startsWith(path) ? 'active' : ''} href={path}>",
          "{label}",
          "</Link>",
          "</li>",
          "))}",
          "</ul>",
          "</div>",
          "<div className="flex-none">",
          "<WalletMultiButton />",
          "</div>",
          "</div>",
          "<div className="flex-grow">{children}</div>",
          "<footer className="footer footer-center p-4 bg-base-300 text-base-content">",
          "<aside>",
          "<p>Generated with @solana-developers/preset-next@0.0.1</p>",
          "</aside>",
          "</footer>",
          "</div>",
          ")",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/app-layout.tsx",
      },
      "global.css": {
        "content": [
          "html,",
          "body {",
          "height: 100%;",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/global.css",
      },
      "layout.tsx": {
        "content": [
          "import './global.css'",
          "import { AppLayout } from './app-layout'",
          "import { SolanaProvider } from './solana-provider'",
          "export const metadata = {",
          "title: 'Welcome to test',",
          "description: 'Generated by @solana-developers/preset-next@0.0.1',",
          "}",
          "export default function RootLayout({ children }: { children: React.ReactNode }) {",
          "return (",
          "<html lang="en">",
          "<body>",
          "<SolanaProvider>",
          "<AppLayout>",
          "{children}",
          "</AppLayout>",
          "</SolanaProvider>",
          "</body>",
          "</html>",
          ")",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/layout.tsx",
      },
      "page.tsx": {
        "content": [
          "export default async function Index() {",
          "return (",
          "<div>",
          "<div>Index</div>",
          "</div>",
          ")",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/page.tsx",
      },
      "solana-provider.tsx": {
        "content": [
          ""use client"",
          "import { WalletError } from '@solana/wallet-adapter-base';",
          "import {",
          "ConnectionProvider,",
          "WalletProvider,",
          "} from '@solana/wallet-adapter-react';",
          "import { WalletModalProvider } from '@solana/wallet-adapter-react-ui';",
          "import { SolflareWalletAdapter } from '@solana/wallet-adapter-solflare';",
          "import { UnsafeBurnerWalletAdapter } from '@solana/wallet-adapter-unsafe-burner';",
          "import { Cluster, clusterApiUrl } from '@solana/web3.js';",
          "import {",
          "createContext,",
          "ReactNode,",
          "useCallback,",
          "useContext,",
          "useMemo,",
          "useState,",
          "} from 'react';",
          "require('@solana/wallet-adapter-react-ui/styles.css');",
          "export interface SolanaProviderContext {",
          "cluster: Cluster;",
          "endpoint: string;",
          "setCluster?: (cluster: Cluster) => void;",
          "}",
          "const Context = createContext<SolanaProviderContext>(",
          "{} as SolanaProviderContext",
          ");",
          "export function SolanaProvider({ children }: { children: ReactNode }) {",
          "const [cluster, setCluster] = useState<Cluster>('devnet');",
          "const endpoint = useMemo(() => clusterApiUrl(cluster), [cluster]);",
          "const wallets = useMemo(",
          "() => [new UnsafeBurnerWalletAdapter(), new SolflareWalletAdapter()],",
          "[cluster]",
          ");",
          "const onError = useCallback((error: WalletError) => {",
          "console.error(error);",
          "}, []);",
          "const value: SolanaProviderContext = {",
          "cluster,",
          "endpoint,",
          "setCluster,",
          "};",
          "return (",
          "<ConnectionProvider endpoint={endpoint}>",
          "<WalletProvider wallets={wallets} onError={onError} autoConnect={true}>",
          "<WalletModalProvider>",
          "<Context.Provider value={value}>{children}</Context.Provider>",
          "</WalletModalProvider>",
          "</WalletProvider>",
          "</ConnectionProvider>",
          ");",
          "}",
          "export function useSolana() {",
          "return useContext(Context);",
          "}",
        ],
        "isBinary": false,
        "path": "test/app/solana-provider.tsx",
      },
    },
    "path": "test/app",
  },
}
`;
